cmd/keytransparency-client/cmd/hammer.go

```diff
@@ -0,0 +1,219 @@
+	var wg sync.WaitGroup
+	go recordLatencies(ctx, times)
+	go generateJobs(ctx, jobs)
+
+	// Slowly add workers up to maxWorkers
+	ramp := time.NewTicker(time.Duration(ramp.Nanoseconds() / int64(maxWorkers)))
+	for ; workers < maxWorkers && ctx.Err() == nil; <-ramp.C {
+		workers++
+		wg.Add(1)
+		go runJobsInThread(ctx, workers, jobs, times, &wg)
+	}
+	ramp.Stop()
+	wg.Wait()
+}
+
+type job func() error
+
+func generateJobs(ctx context.Context, jobs chan<- job) {
+	i := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case jobs <- bindJob(ctx, i):
+			i++
+		}
+	}
+}
+
+func bindJob(ctx context.Context, i int) func() error {
+	userID := fmt.Sprintf("user_%v", i)
+	return func() error {
+		return writeOp(ctx, "app1", userID)
+	}
+}
+
+func runJobsInThread(ctx context.Context, id int, jobs <-chan job, times chan<- time.Duration, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case f := <-jobs:
+			start := time.Now()
+			if err := f(); err != nil {
+				log.Printf("f(): %v", err)
+				continue
+			}
+			times <- time.Since(start)
+		}
+	}
+}
+
+func recordLatencies(ctx context.Context, times <-chan time.Duration) {
+	//start := time.Now()
+	latencies := make([]float64, 0, 1000)
+	refresh := time.NewTicker(250 * time.Millisecond)
+	newworker := time.NewTicker(time.Duration(ramp.Nanoseconds() / int64(maxWorkers)))
+	qps := ratecounter.NewRateCounter(5 * time.Second)
+
+	qpsData := new(tm.DataTable)
+	qpsData.AddColumn("Workers")
+	qpsData.AddColumn("QPS")
+	qpsData.AddRow(0, 0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case t := <-times:
+			latencies = append(latencies, t.Seconds())
+			qps.Incr(1)
+		case <-refresh.C:
+			draw(latencies, qpsData, qps.Rate())
+		case <-newworker.C:
+			qpsData.AddRow(float64(workers), float64(qps.Rate()))
+		}
+	}
+}
+
+func draw(latencies []float64, data *tm.DataTable, qps int64) {
+	tm.Clear()
+	tm.MoveCursor(0, 0)
+
+	// Global stats
+	tm.Printf("Workers: %v\n", workers)
+	tm.Printf("Total Requests: %v\n", len(latencies))
+	tm.Printf("Recent QPS:    %v\n", qps)
+
+	// Threads per QPS graph
+	tm.Printf("QPS Chart:\n")
+	chart := tm.NewLineChart(100, 20)
+	tm.Println(chart.Draw(data))
+
+	// Latency Hist
+	tm.Printf("Latency Histogram:\n")
+	width := 50
+	height := 10
+
+	hist := histogram.Hist(height, latencies)
+	histogram.Fprint(tm.Output, hist, histogram.Linear(width))
+
+	// Console output
+	box := tm.NewBox(180, 10, 0)
+	box.Write(memLog.Bytes())
+	tm.Println(box)
+
+	tm.Flush()
+}
+
+// writeOp performs one write command and returns the time it took to complete.
+func writeOp(ctx context.Context, appID, userID string) error {
+	timeout := viper.GetDuration("timeout")
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	userCreds := authentication.GetFakeCredential(userID)
+	c, err := GetClient(ctx, userCreds)
+	if err != nil {
+		return fmt.Errorf("error connecting: %v", err)
+	}
+
+	// Update.
```

Comment:
Comment is not adding much value. Suggest either expanding or removing it.

Reply:
done